docs/examples/sam_mark_duplicates.md

```diff
@@ -0,0 +1,186 @@
+        for op, length in cigar_ops:
+            if op in ["M", "=", "X", "D", "N"]:
+                aligned_length += length
+
+        # Find trailing soft clips (S operations at the end)
+        for op, length in reversed(cigar_ops):
+            if op == "S":
+                trailing_soft_clips += length
+            else:
+                break  # Stop at first non-S operation from the end
+
+        return pos + aligned_length + trailing_soft_clips - 1
+    
+df = df.with_columns(
+    pl.struct(["pos", "cigar", "flag"])
+    .map_elements(get_unclipped_5_prime_start_position)
+    .alias("unclipped_5p_start_pos")
+)
+
+df.head()
+```
+
+## Group the reads into pairs
+We assume that the qname corresponds to read pairs. We'll also be sure to include the information needed to deduplicate read pairs:
+- Reference genome name
+- 5' start positions
+- Strand flags
+- Quality scores
+
+```{code-cell} ipython3
+pairs_df = df.group_by("qname").agg(
+    [
+        pl.col("rname").alias("rnames"),
+        pl.col("unclipped_5p_start_pos").alias("5p_positions"),
+        ((pl.col("flag") & STRAND_BIT) != 0).alias("strands"),
+        pl.col("qual").alias("quals"),
+    ]
+)
+pairs_df.head()
+```
+
+## Process pair reads
+
+Build a deduplication key and sum the quality scores for each pair (for resolution in the next step)
+
+```{code-cell} ipython3
+def get_quality_score_sum(qual_str):
+    """Calculate the sum of quality scores from a string of quality scores"""
+    return sum(ord(c) - 33 for c in qual_str if c != " ")
+
+
+def build_dedup_key(rnames, positions, strands):
+    """Makes a dedup key for a read pair"""
+    items = sorted(zip(rnames, positions, strands))
+    if len(items) < 2:
+        print(f"WARNING: read is missing pair: {items}")
+        return None
+    return f"{items[0][0]}:{items[0][1]}:{items[0][2]}__{items[1][0]}:{items[1][1]}:{items[1][2]}"
+
+
+pairs_df = pairs_df.with_columns(
+    pl.struct(["rnames", "5p_positions", "strands"])
+    .map_elements(
+        lambda s: build_dedup_key(s["rnames"], s["5p_positions"], s["strands"]),
+        return_dtype=pl.String
+    )
+    .alias("dedup_key"),
+    pl.col("quals")
+    .map_elements(
+        lambda qlist: sum(get_quality_score_sum(q) for q in qlist), 
+        return_dtype=pl.Int64
+    )
+    .alias("total_quality"),
+).filter(pl.col("dedup_key").is_not_null())
+
+pairs_df[('dedup_key', 'total_quality')].head()
+```
+
+## Count the duplicates
+
+We choose the best read pair across duplicates by the best quality score total
+
+```{code-cell} ipython3
+# Resolve duplicate pairs by sorting by total quality and taking the best pair
+best_pairs_df = pairs_df.sort("total_quality", descending=True).unique(
+    subset=["dedup_key"]
+)
+
+# Get the total number of duplicates
+total_pair_dups = pairs_df.height - best_pairs_df.height
+
+print(
+    best_pairs_df.head(),
+    "\nTotal pair duplicates:",
+    total_pair_dups,
+)
+```
```

Comment:
Left the output as just the total count - should I add any other summary stats? Maybe mark duplicates in the original df's flag col and show the distribution of duplicates?